### ISSUE TITLE
Suppressed Panics

### ISSUE THREAD

**Issue by radeusgd** [enso-org/enso#6440]
We currently have at least 3 places where we run some 'finalization' logic:
- `Managed_Resource`
- `Panic.with_finalizer` (essentially a `finally` clause)
- `Connection.run_within_transaction` (rolling back if a panic is thrown)

In each of these places, the finalization logic is triggered when handling a Panic thrown within some inner action, and that panic is supposed to be propagated further after finalization is completed. 

However, the finalization logic may throw the panic too.

Currently, most of the time we decide that the panic from the finalizer takes precedence and it is propagated instead of the original panic. But in this case we lose information about this original panic.

We may want to develop a way to somehow retain information about both panics. We can propagate only one, but we can, for example attach additional information to it.

Java has the concept of [Suppressed Exceptions](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Throwable.html#getSuppressed()) serving a similar purpose. We could get inspired by it.

One issue is that the payload of a Panic can be of any type, so we cannot reliable enhance it with additional data. However, apart from the payload, the panic has additional _metadata_ (like the stack-trace), accessible through the `Caught_Panic` interface. We may be able to exploit that to also provide this additional information.

- [ ] Figure a way to attach suppressed exception information to panics.
	- Probably much like stack trace - add the metadata and make it accessible through `Caught_Panic`.
	- Have a method allowing to attach the suppressed panics to another `Caught_Panic`.
- [ ] Modify the existing mechanisms (destructor of `Managed_Resource`, `Panic.with_finalizer`, `Connection.run_within_transaction`) to still propagate the exception from the destructor, but attach the original exception as 'suppressed'.
- [ ] Add tests for the amended behaviour.
- [ ] Possibly, figure out how we should display this information in the Test Suite and IDE.

**Comment by JaroslavTulach** [comment]
The standard way to attach additional information to objects in Enso are `Warnings`. Can't we just attach the second `Panic` to the `payload` of the first as a warning? Then the information remains there and it shall be transparent for those who don't care.

> Nice! I think it's a very good idea.

If this `Warning` idea could fly, then there is nothing to change on the engine side. All the changes need to be handled in _stdlib_

**Comment by radeusgd** [comment]
> The standard way to attach additional information to objects in Enso are `Warnings`. Can't we just attach the second `Panic` to the `payload` of the first as a warning? Then the information remains there and it shall be transparent for those who don't care.

Nice! I think it's a very good idea.

It won't work once the panic is propagated and caught at Java side, but that will probably be rare so I guess not a big issue.

As long as we ensure that warnings are correctly to preserved when a panic is being thrown and caught (i.e. if I throw a payload with warnings and then catch it, I see the warnings), I think this could be a great solution to this - we will not have to introduce yet another additional concept.

TODO: test if the warnings are now presreved when throwing and accessible when catching

**Comment by jdunkerley** [comment]
@radeusgd - please review and lets work out if we should push this forward.

**Comment by radeusgd** [comment]
> The standard way to attach additional information to objects in Enso are `Warnings`. Can't we just attach the second `Panic` to the `payload` of the first as a warning? Then the information remains there and it shall be transparent for those who don't care.
>
> If this `Warning` idea could fly, then there is nothing to change on the engine side. All the changes need to be handled in _stdlib_

So I have checked and the warnings are lost after the panic is thrown and caught:
```
from Standard.Base import all

type My_Error
    Value x

type My_Warning
    Value x

main =
    foo x =
        p = Warning.attach (My_Warning.Value "warning") (My_Error.Value "error")
        IO.println p
        IO.println (Warning.get_all p)
        if x == 0 then Panic.throw p else x
    
    Panic.catch Any (foo 0) caught_panic->
        IO.println caught_panic.payload
        IO.println (Warning.get_all caught_panic.payload)
```
gives
```
(My_Error.Value 'error')
[Value warning]
(My_Error.Value 'error')
[]
```

----

Moreover, after some further consideration, I don't think that overloading the meaning of warnings for another usage is a good idea - it's essentially _ab_using warnings to get a new feature done at low cost.

I think instead, we should extend the `PanicException` to allow to contain a field for suppressed panics, then such suppressed panics could be read through the `Caught_Panic` interface.

How to add a suppressed panic? I guess we could have an operation `Caught_Panic.add_suppressed : Any -> Caught_Panic` that creates a modified instance and can later be rethrown.

---

Having said that, I don't think this is a priority feature - having the 3 constructs mentioned above that 'may' suppress panics, I haven't encountered a situation where I had problems debugging an error because of lack of this suppressed panic info - in practice very rarely panics occur within such 'cleanup' phases, and if they do, we can usually manually add some logging statements.

So I think we might put this feature on hold and look into it once it actually shows up in practical scenarios.

**Comment by JaroslavTulach** [comment]
> I haven't encountered a situation where I had problems debugging an error because of lack of this suppressed panic info
> in practice very rarely panics occur within such 'cleanup' phases, and if they do, we can usually manually add some logging statements.

Yeah, just add logging statement or report a bug describing the real problem to help us evaluate each issue on a case by case basis. Closing this general issue.
